### PR TITLE
feat: get function name from imports

### PR DESCRIPTION
**Pull request by antonk52**
Awesome plugin!

Many teams use naming conventions for libraries like `classnames` and I think this is something that can be nicely handled by this plugin. I've added a look up feature as well a test for it. I also touched up the previous test to include imports. I am open up for suggestions. 

Cheers

### CODE REVIEW

File: src/index.js
Comment (merceyz) [inline src/index.js]:
```suggestion
          names = ['clsx', 'classNames']
```
It's not guaranteed that the file contains imports, so it should always contain the defaults.

File: test/objects.test.js
Comment (merceyz) [inline test/objects.test.js]: Please revert all changes in this file, the tests here shouldn't check this.

File: src/index.js
Comment (merceyz) [inline src/index.js:16]: Do you want to extend this to handle `const cx = require('classnames')`?

**Review by merceyz — CHANGES_REQUESTED** [review]
Thank you for contributing!

Really like the way you did this, my plan was to use settings so that users could specify what they use, but this is much better :+1:.

I left some comments inline, please take a look.

File: src/index.js
Comment (antonk52) [reply src/index.js]: Thank you for your feedback, defaults sound like a good idea yet these may conflict with other libraries or variables plugin users may use. Perhaps these can be the defaults which users specify in plugin settings. What do you think?

File: test/objects.test.js
Comment (antonk52) [reply test/objects.test.js]: As of right now the plugin will not convert the usage of `classnames` or `clsx` if there are no imports. That's the reason why I had to add it.

File: src/index.js
Comment (antonk52) [reply src/index.js:16]: Great idea. It's a little odd to use `require` nowadays in the client side code but it is something that should be handled too. Thanks

File: src/index.js
Comment (merceyz) [reply src/index.js]: > Perhaps these can be the defaults which users specify in plugin settings. What do you think?

A setting that overrides the defaults would be the way to go.

Comment (antonk52) [reply src/index.js]: Definitely gonna add the user options, yet not sure about `defaultNames` as these still may conflict with other function names. I believe a better option is to not to modify code if we are not sure whether it is `classnames` or `clsx`.

**Comment by antonk52** [comment]
I've added the `defaultNames` plugin option, handling for `require` and tests for both. Still not sure if default options from the plugin itself is something that should be implemented. Perhaps the phrasing for options description has room for improvement 😅

File: src/index.js
Comment (merceyz) [reply src/index.js]: > yet not sure about defaultNames as these still may conflict with other function names

Then i'd rather have the user disable it, the plugin should do it's job out of the box.
If they have some setup that causes issues with this, they can disable it.

**Comment by merceyz** [comment]
Thank you, I have some other changes that I want to get in before this so i'll leave this as is for now.
Once ready, i'll make some final changes and get this merged :+1:

**Comment by antonk52** [comment]
I would like to ask you to reconsider adding "classNames" and "clsx" as defaults unless users specify otherwise. These are no doubt popular names to import `classnames` and `clsx` libs yet once my PR is merged, those will no longer be needed as defaults since in the most cases people use these helpers in the same file as the lib is imported. This makes the default name not only superfluous but it may also conflict with other helpers plugin users may have. The proposed feature of looking up name in the same file should cover the most cases. If we cannot make sure that the function call is evoking `classnames` or `clsx` I believe it is a safer option to leave it alone than to leave a possibility breaking code unintentionally. Please keep in mind once someone gets this kind of bug it will not be straight forward to trouble shoot. this is why I am asking one more time to reconsider the default values. Either way thanks again for the plugin 👍 

**Update**
I have updated my branch with current master and most tests fail since the test structure has been redone and the imports are missing. Perhaps we can discuss adding library imports dynamically in the tests since the amount of tests is probably going to grow in the future and changes to boilerplate are going to be more time consuming.

**Comment by merceyz** [comment]
I agree with you(always did), but didn't want a breaking change so early on.
However the bugs that can come from just looking for function names outweigh that a lot.

The way I see it, this is how it should go:
Default to look for `clsx` and `classnames` import/require and use the names found there, which users should be able to override.
There will also be an option for specifying function names to look for directly.

This PR is currently blocked by https://github.com/babel-utils/babel-plugin-tester/issues/40

**Comment by antonk52** [comment]
Oh, thank you for the clarification. My bad I must've misinterpreted earlier.

> However the bugs that can come from just looking for function names outweigh that a lot.

Can you please come up with an example or two. Not sure I get this.

> which users should be able to override

And this one too. Do you mean reassigning it to another variable name or using it within another function?

**Comment by merceyz** [comment]
> Can you please come up with an example or two. Not sure I get this.

I mean that you're right, the issues that can come from just matching on `clsx(...)` outweighs my wish for not adding a breaking change, it should use imports like you suggested

> And this one too. Do you mean reassigning it to another variable name or using it within another function?

For example if the user has their own version of `classnames` called `my-classnames`, they can specify that it should look for that import as well.

Example:
```javascript
const options = {
  plugins: [
    [
      "babel-plugin-optimize-clsx",
      {
        // Matches all functions called "foo"
        functionNames: ["foo"],
        // clsx and classnames are default
        libraries: ["clsx", "classnames", "my-classnames"]
      }
    ]
  ]
};

```

**Review by merceyz — APPROVED** [review]

**Comment by merceyz** [comment]
@antonk52 I've made some changes, please take a look and let me know what you think.

File: src/options.js
Comment (antonk52) [inline src/options.js]: Do we really need lodash here instead of native `Object.assign`?

**Review by antonk52 — COMMENTED** [review]
I only have one concern about lodash usage, every thing else looks great! Look forward for this to get merged and start using it in my projects

File: src/options.js
Comment (merceyz) [reply src/options.js]: Not at all, changed it to use spread operator

**Comment by merceyz** [comment]
> Update
I have updated my branch with current master and most tests fail since the test structure has been redone and the imports are missing. Perhaps we can discuss adding library imports dynamically in the tests since the amount of tests is probably going to grow in the future and changes to boilerplate are going to be more time consuming.

Didn't see this until now, but took care of that with the feature I added to `babel-plugin-tester`

> I only have one concern about lodash usage

Yeah, that was a bit excessive on my part

**Comment by antonk52** [comment]
Is there something blocking this PR atm?

**Comment by merceyz** [comment]
> Is there something blocking this PR atm?

Nope, thank you :+1:

**Pull request merged**
